docs/sources/mimir/references/architecture/mimir-query-engine.md

```diff
@@ -0,0 +1,101 @@
+---
+description: The Mimir query engine (MQE) is responsible for evaluating PromQL queries.
+menuTitle: Mimir query engine
+title: Grafana Mimir query engine
+weight: 100
+---
+
+The Mimir query engine (MQE) is an experimental alternative to Prometheus' query engine.
+It is used in [queriers]({{< relref "./components/querier" >}}) to evaluate PromQL queries.
+
+MQE produces equivalent results to Prometheus' engine, generally uses less memory and CPU
+than Prometheus' engine, and evaluates queries at least as fast, if not faster.
+It supports almost all stable PromQL features, and transparently falls back to Prometheus'
+engine for queries that use unsupported features.
+
+## How to enable MQE
+
+MQE is experimental and disabled by default. It can be enabled by setting the
+`-querier.query-engine=mimir` CLI flag on queriers, or setting the equivalent YAML
+configuration file option.
+
+## Fallback to Prometheus' engine
+
+By default, MQE will fall back to Prometheus' engine for any queries that use unsupported
+features.
+
+This behaviour can be disabled by setting the `-querier.enable-query-engine-fallback=false`
+CLI flag on queriers, or setting the equivalent YAML configuration file option. If fallback
+is disabled and MQE receives a query it does not support, then the query will fail.
+
+You can force a query supported by MQE to use Prometheus' engine by adding the
+`X-Mimir-Force-Prometheus-Engine: true` HTTP header to the query request. This header only
+has an effect if fallback is enabled.
+
+## Query memory consumption limit
+
+MQE supports enforcing a per-query memory consumption limit. This allows you to ensure that
+a single memory-hungry query cannot monopolise a large proportion of available memory in a
+querier, or cause it to exhaust all available memory and crash.
+
+While evaluating a query, MQE estimates the memory consumed by the query, such as memory used
+for the final result and any intermediate calculations, and aborts the query with an
+[`err-mimir-max-estimated-memory-consumption-per-query`]({{< relref "../../manage/mimir-runbooks#err-mimir-max-estimated-memory-consumption-per-query" >}})
+error if the estimate exceeds the configured limit.
+
+The estimate is based on the memory consumed by samples currently held in memory for query
+evaluation. This includes both raw samples decoded from chunks and samples held in memory as
+intermediate results of calculations, or the final result. It also includes some other large
```

Comment:
`This includes both raw samples decoded from chunks and samples held in memory as intermediate results of calculations, or the final result.`

This sentence reads somewhat confusing to me. Does it mean samples held in memory has either intermediate or final results?